### ISSUE TITLE
Add "X-CF-InstanceIndex" header to gorouter metrics

### ISSUE THREAD

**Issue by antonsoroko** [cloudfoundry/gorouter#163]
Hi there!

It would be very nice to have "X-CF-InstanceIndex" header so that `InstanceIndex` field of songe-go's HttpStartStop struct will have real value.

https://github.com/cloudfoundry/sonde-go/blob/master/events/http.pb.go#L230
https://github.com/cloudfoundry/dropsonde/blob/master/factories/factories.go#L40

Right now `InstanceIndex` is always `nil`, so there is no way to understand which instance of the application is associated with a particular `HttpStartStop` event.

But in my humble opinion this information is pretty useful (for logs and monitoring, for example).

I hope that @Nino-K remembers discussion of this question in `#routing` slack channel :-)

Could you please consider adding this header to the gorouter?

Thank you in advance!

**Comment by shashwathi** [comment]
Hi @antonsoroko : 

Thank you for raising this issue. Currently we set `X-CF-InstanceID` and `X-CF-ApplicationID` on the request and sonde-go should use these values on HTTP Start Stop events. Is this information not enough for monitoring? 

We have a [story](https://www.pivotaltracker.com/story/show/127324153) in our backlog to add instance ID to access logs.

Regards
Shash

**Comment by antonsoroko** [comment]
@shashwathi 
Hi.
Thank you for your response.

`X-CF-ApplicationID` is useful, becuuse there is the way to convert this UUID to app's name, but `X-CF-InstanceID` is not, because it stores some weird UUID.
`X-CF-InstanceID` can be useful only if you know a way how to convert UUID provided by `X-CF-InstanceID` to simple integer (index of app's instance).
I asked this question in slack, but I'm afraid there is no way to do it :-(

**Comment by antonsoroko** [comment]
Diego, for example, provides human-readable index for app's instance:
https://github.com/cloudfoundry/sonde-go/blob/master/events/metric.pb.go#L92

**Comment by shashwathi** [comment]
@antonsoroko 

Thank you for your response. This is a reasonable request, tagging PM's on this thread

@shalako @abbyachau

Regards
Shash

**Comment by antonsoroko** [comment]
@shashwathi @swetharepakula 
Hi!
Any news?)

**Comment by shashwathi** [comment]
@shalako @abbyachau 

Any update on when this story will be prioritized?

**Comment by shalako** [comment]
This thread only mentions implementation details. Before prioritizing, I would like to see the user-facing acceptance criteria. Who (admin/app dev) needs the app index, and from where? The gorouter already logs the app index in the access log, so an app dev would see it in the response for `cf logs myapp`. Are we talking about an admin using `cf nozzle` to connect to the firehose?

The story mentions an http header, but I don't understand why this would be useful to the application receiving the request. Is this not already available to the app in a runtime env var?

I believe Shash and I discussed this some time ago but I don't recall the answer.

**Comment by antonsoroko** [comment]
@shalako 
Hi!

>I would like to see the user-facing acceptance criteria

My bad, sorry. Let me explain:

The main goal of this feature is to provide `App Instance Index` to firehose nozzles (e.g. influx/graphite/opentsdb/etc), so nozzle developers will be able to include this information into app's metrics.

And I'm talking not about realtime usage (like in case with `cf logs`/`cf nozzle`) but about long-term storing of metrics with useful names (name which contains not only app's name, but also app's index, to be able to differentiate different instances.)

Without this `InstanceID` all HTTP metrics are flattened - there is no way to determine which instance of application is causing problems (e.g. 50x errors) or which instance is more loaded.

If I poorly described the idea, please feel free to ask  clarifying questions.

**Comment by shalako** [comment]
It appears that two `HttpStartStop` events are emitted for each request to an app. @shashwathi, why is that? Only one of them includes the `applicationId` key, which has two values; what does `low` and `high` mean? (confusing). 

```
origin:"gorouter" eventType:HttpStartStop timestamp:1489684987782380442 deployment:"cf" job:"router" index:"62288e42-d572-49ed-bb61-4924cfef4595" ip:"10.0.32.6" httpStartStop:<startTimestamp:1489684987778084060 stopTimestamp:1489684987782368894 requestId:<low:13136963336295494592 high:14507952726188751970 > peerType:Client method:GET uri:"http://dora-scoen.superman.cf-app.com/" remoteAddress:"130.211.3.26:63677" userAgent:"curl/7.51.0" statusCode:200 contentLength:13 applicationId:<low:4921262785486289610 high:7221885839010293168 > instanceId:"7e21fcd6-2320-4e99-681f-0f02d3603fbd" forwarded:"209.234.137.222" forwarded:"35.186.227.42" forwarded:"130.211.3.26" >

origin:"gorouter" eventType:HttpStartStop timestamp:1489684987782600281 deployment:"cf" job:"router" index:"62288e42-d572-49ed-bb61-4924cfef4595" ip:"10.0.32.6" httpStartStop:<startTimestamp:1489684987777892046 stopTimestamp:1489684987782591181 requestId:<low:17096107401131619502 high:2063030011441055052 > peerType:Server method:GET uri:"http://dora-scoen.superman.cf-app.com/" remoteAddress:"130.211.3.26:63677" userAgent:"curl/7.51.0" statusCode:200 contentLength:13 forwarded:"209.234.137.222" forwarded:"35.186.227.42" >
```

We also emit a metrics event for the access log, which does include `app_index`:
```
origin:"gorouter" eventType:LogMessage timestamp:1489684987782720620 deployment:"cf" job:"router" index:"62288e42-d572-49ed-bb61-4924cfef4595" ip:"10.0.32.6" logMessage:<message:"dora-scoen.superman.cf-app.com - [2017-03-16T17:23:07.777+0000] \"GET / HTTP/1.1\" 200 0 13 \"-\" \"curl/7.51.0\" \"130.211.3.26:63677\" \"10.0.16.10:60024\" x_forwarded_for:\"209.234.137.222, 35.186.227.42\" x_forwarded_proto:\"http\" vcap_request_id:\"c033393f-8dde-4fb6-62fc-a356239a56c9\" response_time:0.004684844 app_id:\"ca7a42ef-6dd6-4b44-b0e5-a14ae84a3964\" app_index:\"0\"\n" message_type:OUT timestamp:1489684987782719501 app_id:"ca7a42ef-6dd6-4b44-b0e5-a14ae84a3964" source_type:"RTR" source_instance:"1" >
```

This matches the access log we write locally and stream to external syslog consumers.
```
2017-03-16T10:23:07.78-0700 [RTR/1]      OUT dora-scoen.superman.cf-app.com - [2017-03-16T17:23:07.777+0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.51.0" "130.211.3.26:63677" "10.0.16.10:60024" x_forwarded_for:"209.234.137.222, 35.186.227.42" x_forwarded_proto:"http" vcap_request_id:"c033393f-8dde-4fb6-62fc-a356239a56c9" response_time:0.004684844 app_id:"ca7a42ef-6dd6-4b44-b0e5-a14ae84a3964" app_index:"0"
```

Sounds like @antonsoroko would like to see something like `applicationInstanceIndex` or `app_index` in the `HttpStartStop` event, set to the same value as we log in the access log.

**Comment by antonsoroko** [comment]
>Sounds like @antonsoroko would like to see something like applicationInstanceIndex or app_index in the HttpStartStop event, set to the same value as we log in the access log.

Yes, that's exactly what I want. :-)

And as I mentioned in the first message, `sonde-go` and `dropsonde` libraries already sends `InstanceIndex` field in HTTP metric, they just don't have appropriate information to fill this field (so this field is always `nil`). As you can see in code, `dropsonde` expects "X-CF-InstanceIndex" header and if it exists `dropsonde` will fill `InstanceIndex` field with value of "X-CF-InstanceIndex" header.

Actually, it looks like that once upon a time gorouter had this header. Or at least someone discussed this header and then implemented it only at `dropsonde` side. But I could be wrong, of course.

**Comment by shalako** [comment]
As there is no known product value in sending this data as a header, that may be an unfortunate user-facing side-effect of delivering the behavior you've requested: that the app index key is available in the `HttpStartStop` metric from the firehose. I would not have the team send the header if we could avoid it.

**Comment by antonsoroko** [comment]
@shalako 
May I ask you to discuss possible options of implementing this feature with your team, please?
Thanks in advance!

**Comment by shalako** [comment]
Prioritizing

**Comment by antonsoroko** [comment]
Hi @shalako .
Any news?)

**Comment by shashwathi** [comment]
@antonsoroko : 

Sorry for the delay in responding to this issue. This is fixed in routing release [v0.157.0](https://github.com/cloudfoundry-incubator/routing-release/releases/tag/0.157.0). Could you verify if this fix actually works on your deployment?

Regards
Shash

**Comment by shashwathi** [comment]
@antonsoroko : 

I am going to close this issue for now. Feel free to open another issue if you are not able to see this metric in your deployment.

Regards
Shash
